Feat/auth part 2

philipeatela opened the pull request:
Este PR inclui:

- Middleware de validacao que permite facilmente validar rotas da API simplesmente passando o DTO na hora de definir a rota. As rotas validadas vao retornar 400 em erro. Precisei modificar um pouco dos DTOs encaixar no middleware.
- Pacote de logs para facilitar debugar e registrar erros
- Criado uma classe de erro base & tipos de erro que extendem a base, para ter uma excecao especifica para cada status code. 
- Finalizado endpoints de cadastro & login, retornando os códigos de erro corretos (401 para credenciais invalidas, 400 para payload invalido, 404 para usuario não encontrado, 500 caso de algum erro no banco)
- Testes de unidade para os servicos de auth e o que faltava do user

Commit:
chore: fix docker compose command syntax, add tslog dependency

Commit:
feat: auth error handling and logging

Commit:
feat: validation middleware & login endpoint validation

Commit:
feat: validation on signup endpoint

Commit:
chore: swap libraries to fix build

Commit:
fix: jest regex to match js files only

Commit:
test: added unit tests for auth & user service

philipeatela commented on puncto-back/jest.config.js:
```diff
@@ -1,4 +1,7 @@
   testEnvironment: 'node',
+  "testMatch": [
+    "**/?(*.)+(spec|test).js?(x)",
+  ],
```
os testes estavam executando duas vezes, uma pros arquivos `.ts` e uma pros `.js`. Modifiquei para executar só o dos arquivos `.js` (a ideia é que sempre antes de executar os testes os arquivos `ts` viram `js`, então não precisa executar os test.ts).

gabrielchaves7 replied:
Essa configuração atrapalha de debugar os testes usando o plugin que eu especifiquei no readme, com aquele plugin você consegue debugar o teste escrito em typescript, e pra isso ele precisa rodar os dois.

philipeatela commented on puncto-back/package.json:
```diff
@@ -20,7 +20,7 @@
     "build": "npm run clean && tsc",
     "test": "npm run build && jest",
-    "mongodb": "docker compose up -d",
+    "mongodb": "docker-compose up -d",
```
corrigido a sintaxe desse comando

philipeatela commented on puncto-back/src/api/user.ts:
```diff
@@ -1,56 +1,63 @@
 // @TODO validar email unico antes de cadastrar
-// @TODO validacao de erros
-router.post('/signup', async (req: Request, res: Response) => {
+router.post('/signup', validate(UserDto), async (req: Request, res: Response) => {
```
É so passar `validate(DTO)` como o segundo parametro da funcao de definicao da rota, que o middleware cuida do resto e valida todos os payloads recebidos no endpoint de acordo com o DTO recebido

philipeatela commented on puncto-back/src/dto/loginDto.ts:
```diff
@@ -1,14 +1,14 @@
+    this.email = '';
+    this.password = '';
   }
 }
```
@gabrielchaves7 achei que estava meio redundante definir a interface e a classe com os mesmos campos, então fiz essa modificacao aqui.

Isso muda um pouco como os DTOs são utilizados no arquivo de teste por exemplo, mas acho que ficou melhor. De uma olhada e me avise o que vc acha.

Adicionei esse pacote `class-validator` tbm, que da esses decorators que podemos usar para validar os campos. O middleware de validacao depende dele, então se quiser utilizar nas rotas é necessário definir essas regras no DTO.

gabrielchaves7 replied:
Eu também não tinha visto muito sentido nas interfaces, só fiz a pra company pra manter o padrão que você estava fazendo. Gostei desses decorators, vou adicionar depois

philipeatela commented on puncto-back/src/exceptions/UnauthorizedException.ts:
```diff
@@ -0,0 +1,10 @@
+        this.message = message;
+        this.type = 'Unauthorized';
+        this.statusCode = 401;
+      }
```
@gabrielchaves7 ve o que vc acha desse esquema de error handling também

Pensei que podemos separar os tipos de erro só pelo codigo (unauthorized, database, not found, etc) mesmo, ao inves de criar um específico pra cada rota

gabrielchaves7 replied:
Esta ótimo

philipeatela commented on puncto-back/src/service/authService.ts:
```diff
@@ -6,14 +6,19 @@ import jwt from 'jsonwebtoken';
+
+    if (!authenticated) {
+      throw new UnauthorizedException('Invalid credentials');
     }
```
Apenas services dão `throw` em erros específicos. Na camada do Controller, vamos só retornar para o cliente o código do erro dado pelo Service

philipeatela commented on puncto-back/src/service/userService.test.ts:
```diff
@@ -1,4 +1,4 @@
-import { UserDto, IUserProps } from '../dto/userDto';
+import { UserDto } from '../dto/userDto';
```
Aqui podemos ver a "nova" forma de usar os DTOs, sem precisar mais daquele helper `getUser`. Me avise do que achar tbm!

gabrielchaves7 committed:
Update jest.config.js

Alterando o jest.config só para manter a funcionalidade de debugar os testes em typescript, depois se você quiser podemos conversar sobre
